rote/minitermite/doc/src2term.1
.\" Man page generated from reStructeredText.
.
.TH MINI-TERMITE 1 "2012-07-10" "1.1" "Compilers and Programming Languages"
.SH NAME
mini-Termite \- Term-based intermediate representation for source code
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" -*- rst -*-
.
.SH SYNOPSIS
.sp
src2term [FRONTEND OPTIONS] [\-\-dot] [\-\-pdf] src1.c src2.cpp src3.f ... [\-o termfile.pl]
.SH DESCRIPTION
.INDENT 0.0
.INDENT 3.5
.sp
Parse one or more source files and convert them into a TERMITE file.
Header files will be included in the term representation.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B Options:
.
[FRONTENT OPTIONS] will be passed to the C/C++/Fortran frontend.
.INDENT 7.0
.TP
.B \-\-rose\-help
.
Display the help for the C/C++/Fortran frontend.
.UNINDENT
.INDENT 7.0
.TP
.BI \-o,  \-\-output \ <termfile.pl>
.
Write the output to <termifile.pl> instead of stdout.
.UNINDENT
.INDENT 7.0
.TP
.B \-\-dot
.
Create a dotty graph of the syntax tree.
.TP
.B \-\-pdf
.
Create a PDF printout of the syntax tree.
.TP
.B \-\-stratego
.
Create term output compatible with the Stratego/XT.
.UNINDENT
.UNINDENT
.sp
This program was built against minitermite 1.1,
please report bugs to <\fI\%adrian@complang.tuwien.ac.at\fP>.
.SH About
.sp
Minitermite is a light\-weight library that transforms C, C++, and
Fortran programs into a term\-based representation of their abstract
syntax trees (AST) and vice versa.
.SH Related work, Provenance
.sp
Minitermite is a spin\-off (friendly fork) of Termite [1] which is
distributed as part of the SATIrE program analysis framework [2] from
the Computer Languages group at Vienna University of Technology,
Austria.
.sp
[1] \fI\%http://www.complang.tuwien.ac.at/adrian/termite\fP
[2] \fI\%http://www.complang.tuwien.ac.at/satire\fP
.SH Authors
.sp
The corresponding author and maintainer is Adrian Prantl. The original
Termite was based on an earlier prototype by Christoph Bonitz and
Markus Schordan.  Gergo Barany contributed the term_lint grammar
checker, as well as support for the ICFG and other PAG\-based analyses
and numerous bug fixes and improvements. Several testcases were
contributed by Viktor Pavlu. The majority of the Clang front\-end
integration was contributed by Dietmar Ebner with some help from
Florian Brandner.
.SH Installation instructions
.sp
The following commands should do the job on a UNIX\-like system:
.sp
.nf
.ft C
mkdir build && cd build
\&../configure \-\-prefix=/usr/local \-\-with\-rosedir=/opt/rose \e
             \-\-with\-boostdir=/usr \-\-with\-swi\-prolog=no
make \-j<N> install
make installcheck
.ft P
.fi
.SH Dependencies
.sp
ROSE: \fI\%http://www.rosecompiler.org/\fP
flex, yacc, boost
.sp
Optional: swi\-prolog, clang, doxygen
.SH License
.sp
Just like SATIrE, minitermite is distributed under a permissive
three\-clause BSD license. See COPYING for details.
termlint.pl is covered by the GPLv3.
.\" Generated by docutils manpage writer.
.\" 
.


rote/minitermite/src/src2term.C
/* -*- C++ -*-
Copyright 2006 Christoph Bonitz <christoph.bonitz@gmail.com>
          2007-2009 Adrian Prantl <adrian@complang.tuwien.ac.at>

 * Purpose: create a TERMITE representation of a given AST
 */

#include <iostream>
#include <fstream>

#include <rose.h>

#include <getopt.h>

// GB (2009-02-25): We want to build c2term without ICFG stuff to avoid
// having to link against libsatire. Also, want to avoid PAG DFI stuff.
#define HAVE_SATIRE_ICFG 0
#undef HAVE_PAG
#include "TermPrinter.h"

using namespace std;
using namespace term;

void usage(const char* me) 
{
  cout << "Usage: " << me
  <<" [FRONTEND OPTIONS] [--dot] [--pdf] src1.c src2.cpp src3.f ... [-o termfile.pl]\n"
  <<"  Parse one or more source files and convert them into a TERMITE file."
       << "\n  Header files will be included in the term representation.\n\n"

       << "Options:\n"
       << "  [FRONTENT OPTIONS] will be passed to the C/C++/Fortran frontend.\n\n" 

       << "  --rose-help\n"
       << "    Display the help for the C/C++/Fortran frontend.\n\n"

       << "  -o, --output <termfile.pl>\n"
       << "    Write the output to <termifile.pl> instead of stdout.\n\n"

       << "  --dot\n"
       << "    Create a dotty graph of the syntax tree.\n\n"

       << "  --pdf\n"
       << "    Create a PDF printout of the syntax tree.\n\n"

       << "  --stratego\n"
       << "    Create term output compatible with the Stratego/XT.\n\n"
#if HAVE_SWI_PROLOG
       << "  --stl-engine\n"
       << "    Do not use SWI-Prolog to generate term output.\n\n"
#endif
       << "This program was built against "<<PACKAGE_STRING<<",\n"
       << "please report bugs to <"<<PACKAGE_BUGREPORT<<">."

       << endl;
}

int main(int argc, char** argv) {
  // Turn off the frontend's warnings; they are distracting in the
  // automated test outputs.
  vector<char*> argv1;
  char warningOpt[] = "-edg:w";
  char includeOpt[] = "-I" ROSE_INCLUDE_DIR;
  argv1.push_back(argv[0]);
  argv1.push_back(warningOpt);
  argv1.push_back(includeOpt);
  for (int i = 1; i < argc; ++i)
    argv1.push_back(argv[i]);
  int argc1 = argc+2;
// Process our own options
  const char* outfile = NULL;
  int dot_flag = 0;
  int pdf_flag = 0;
  int stratego_flag = 0;
#if HAVE_SWI_PROLOG
  int stl_flag = 0;
#else
  int stl_flag = 1;
#endif
  int version_flag = 0;
  int help_flag = 0;
  int rose_help_flag = 0;

  static struct option long_options[] = {
    /* These options set a flag. */
    {"version", no_argument, &version_flag, 1},
    {"help", no_argument, &help_flag, 1},
    {"rose-help", no_argument, &rose_help_flag, 1},
    {"dot", no_argument, &dot_flag, 1},
    {"pdf", no_argument, &pdf_flag, 1},
    {"stratego", no_argument, &stratego_flag, 1},
#if HAVE_SWI_PROLOG
    {"stl-engine", no_argument, &stl_flag, 1},
#endif
    /* These don't */
    {"output", required_argument, 0, 'o'},
    {0, 0, 0, 0}
  };

  // Surely not the most efficient thing to do, but we need to remove
  // those pesky -I... and -D... options to not confuse getopt
  for (int i = 1; i < argc; ++i) {
    if (argv[i][0] == '-') {
      char empty[] = "";
      bool edg_opt = true;
      for (struct option *opt = long_options; opt->name; ++opt) {
	if (argv[i]+1 && string(argv[i]+2) == string(opt->name)) edg_opt = false;
	if (argv[i][1] == 'o') edg_opt = false;
      }
      if (edg_opt)
	argv[i] = empty;
    }
  }

  while (1) {
    /* getopt_long stores the option index here. */
    int option_index = 0;
     
    int c = getopt_long(argc, argv, "o:", long_options, &option_index);
     
    /* Detect the end of the options. */
    if (c == -1)
      break;
     
    switch (c) {
    case 'o': outfile = optarg; break;
    default: ; /* ignore - pass to frontend */
    }
  }
  if (help_flag) { 
    usage(argv[0]); 
    return 0;
  }
  if (rose_help_flag) { 
    argc1 = 2;
    argv1[1] = strdup("--help");
    frontend(argc1,&argv1[0]);
    return 0;
  }
  if (version_flag) { 
    cout << argv[0] << " version " << PACKAGE_VERSION << "\n"; 
    return 0;
  }
  if (optind < argc) {
    //infile = argv[optind];
  } else {
    usage(argv[0]);
    return 1;
  }

  // Run the EDG frontend
  SgProject* project = frontend(argc1,&argv1[0]);

  if (dot_flag) {
    //  Create dot and pdf files
    //  DOT generation (numbering:preoder)
    AstDOTGeneration dotgen;
    dotgen.generateInputFiles(project, AstDOTGeneration::PREORDER);
  }
  if (pdf_flag) {
    //  PDF generation
    AstPDFGeneration pdfgen;
    pdfgen.generateInputFiles(project);
  }

  init_termite(argc, argv);

  // Choose the way to construct terms based on the options
  TermFactory* termFactory;
  if (stratego_flag) {
    yy_use_stratego_filter = true;
    termFactory = new StrategoTermFactory();
  } else 
    if (stl_flag)
      termFactory = new STLTermFactory();
#if HAVE_SWI_PROLOG
    else termFactory = new SWIPLTermFactory();
#endif

  // Create prolog term
  BasicTermPrinter tp(*termFactory);
  tp.traverse(project); // With headers

  Term* genTerm = tp.getTerm();
  
  if (outfile) {
    ofstream ofile(outfile);
    ofile << genTerm->getRepresentation() << "." << endl;
    ofile.close();
  } else cout << genTerm->getRepresentation() << "." << endl;
		
  return 0;
}
